R/indentation_linter.R

```diff
@@ -0,0 +1,163 @@
+#' Check that indentation is consistent
+#'
+#' @param indent Block indentation level, used for multi-line code blocks (`{ ... }`), function calls (`( ... )`) and
+#'   extractions (`[ ... ]`, `[[ ... ]]`).
+#' @param use_hybrid_indent Require a block indent for multi-line function calls if there are only unnamed arguments
+#'   in the first line?
+#'   ```r
+#'   # complies only with use_hybrid_indent = TRUE:
+#'   map(x, f,
+#'     additional_arg = 42
+#'   )
+#'
+#'   # complies only with use_hybrid_indent = FALSE:
+#'   map(x, f,
+#'       additional_arg = 42
+#'   )
+#'
+#'   # always complies
+#'   map(x, f,
+#'       additional_arg = 42)
+#'   ```
+#'
+#' @evalRd rd_tags("indentation_linter")
+#' @seealso [linters] for a complete list of linters available in lintr.
+#' @export
+indentation_linter <- function(indent = 2L, use_hybrid_indent = TRUE) {
+  paren_tokens <- c("OP-LEFT-BRACE", "OP-LEFT-PAREN", "OP-LEFT-BRACKET", "LBB")
+  paren_tokens_right <- c("OP-RIGHT-BRACE", "OP-RIGHT-PAREN", "OP-RIGHT-BRACKET", "OP-RIGHT-BRACKET")
+  infix_tokens <- setdiff(infix_metadata$xml_tag, c("OP-LEFT-BRACE", "OP-COMMA", paren_tokens))
+  no_paren_keywords <- c("ELSE", "REPEAT")
+  keyword_tokens <- c("FUNCTION", "IF", "FOR", "WHILE")
+
+  xp_last_on_line <- "@line1 != following-sibling::*[not(self::COMMENT)][1]/@line1"
+
+  if (use_hybrid_indent) {
+    xp_is_not_hanging <- paste(
+      c(
+        glue::glue(
+          "self::{paren_tokens}/following-sibling::{paren_tokens_right}[@line1 > preceding-sibling::*[1]/@line2]"
+        ),
+        glue::glue("self::*[{xp_and(paste0('not(self::', paren_tokens, ')'))} and {xp_last_on_line}]")
+      ),
+      collapse = " | "
+    )
+  } else {
+    xp_is_not_hanging <- glue::glue("self::*[{xp_last_on_line}]")
+  }
+
+  xp_block_ends <- paste0(
+    "number(",
+    paste(
+      c(
+        glue::glue("self::{paren_tokens}/following-sibling::{paren_tokens_right}/preceding-sibling::*[1]/@line2"),
+        glue::glue("self::*[{xp_and(paste0('not(self::', paren_tokens, ')'))}]/following-sibling::SYMBOL_FUNCTION_CALL/
+                      parent::expr/following-sibling::expr[1]/@line2"),
+        glue::glue("self::*[
+                      {xp_and(paste0('not(self::', paren_tokens, ')'))} and
+                      not(following-sibling::SYMBOL_FUNCTION_CALL)
+                    ]/following-sibling::*[1]/@line2")
+      ),
+      collapse = " | "
+    ),
+    ")"
+  )
+
+  xp_indent_changes <- paste(
+    c(
+      glue::glue("//{paren_tokens}[not(@line1 = following-sibling::expr[
```

Comment:
with all of these XPaths, I do wonder about performance, could you do a quick check on some large files vs. some other default linters to see if that concern is valid?

My first thought is: would it be possible to do a quick sanity check on the input file, like `gregexpr("^\\s+", lines)` and if all of the matches are multiples of 2, we're done already? I guess not, since something like this would pass

```r
x =
    too_far_out()
```

Maybe if everything's a multple of 2 and `diff(indent)` is never more than 2?

Reply:
Let's worry about performance when the functionality is final.
I'm not too worried because most of them don't use `*` or `expr` for entry.
Also, I didn't experience noticable performance problems during my testing of CRAN packages.

Reply:
Note for future: If we run into trouble with performance (looking at tools?), the `self::*` could be transformed as to never select `expr` nodes, at the cost of exploding the XPath with all other allowed XML tags.

Reply:
SG. Also for future reference, per the frequency table here, I guess the biggest concern would be `//OP-LEFT-PAREN`:

https://github.com/r-lib/lintr/issues/1358#issue-1260855389